File: NeetCode/Trees/Diameter_Of_A_Binary_Tree/diameter__of__a__binary__tree.py
from typing import Optional, List

from common.treenode import TreeNode, deserialize, printTree, null


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(root):
            # if the root is None, then it's a empty branch
            if root is None:
                return -1
            left = dfs(root.left)
            right = dfs(root.right)
            self.maximum = max(self.maximum, left + right + 2)
            return 1 + max(left, right)

        self.maximum = 0
        dfs(root)
        return self.maximum


if __name__ == "__main__":
    sol = Solution()
    root = [1, null, 2, 3, 4, 5]
    root = deserialize(root)
    print(sol.diameterOfBinaryTree(root))
    print("Running Solution...")

File: NeetCode/Trees/Diameter_Of_A_Binary_Tree/Diameter_Of_A_Binary_Tree.cpp
// Diameter_Of_A_Binary_Tree.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
private:
    int maxFound;

public:
    int diameterOfBinaryTree(TreeNode *root)
    {
        maxFound = 0;
        dfs(root);
        return maxFound;
    }
    int dfs(TreeNode *node)
    {
        if (!node)
        {
            return 0;
        }
        int left = dfs(node->left);
        int right = dfs(node->right);

        maxFound = max(maxFound, left + right);

        return max(left, right) + 1;
    }
};
